From f33ea51b84a3a9a0e84c7eeb4a7ad143ed97f605 Mon Sep 17 00:00:00 2001
From: Imre Kaloz <kaloz@openwrt.org>
Date: Thu, 16 Mar 2023 19:18:25 +0530
Subject: [PATCH 091/281] OpenWrt: 920-mangle_bootargs.patch

init: add CONFIG_MANGLE_BOOTARGS and disable it by default

Enabling this option renames the bootloader supplied root=
and rootfstype= variables, which might have to be know but
would break the automatisms OpenWrt uses.

Signed-off-by: Imre Kaloz <kaloz@openwrt.org>

Change-Id: I6bdc3c9a9dbf531b7640b9f65eb9c75853303b9c
Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>
---
 init/Kconfig |  9 +++++++++
 init/main.c  | 24 ++++++++++++++++++++++++
 2 files changed, 33 insertions(+)

diff --git a/init/Kconfig b/init/Kconfig
index 93fb1ad88201..4fc578f97cd8 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1830,6 +1830,15 @@ config EMBEDDED
 	  an embedded system so certain expert options are available
 	  for configuration.
 
+config MANGLE_BOOTARGS
+	bool "Rename offending bootargs"
+	depends on EXPERT
+	help
+	  Sometimes the bootloader passed bogus root= and rootfstype=
+	  parameters to the kernel, and while you want to ignore them,
+	  you need to know the values f.e. to support dual firmware
+	  layouts on the flash.
+
 config HAVE_PERF_EVENTS
 	bool
 	help
diff --git a/init/main.c b/init/main.c
index aa21add5f7c5..88dd07e5c7d7 100644
--- a/init/main.c
+++ b/init/main.c
@@ -609,6 +609,29 @@ static inline void setup_nr_cpu_ids(void) { }
 static inline void smp_prepare_cpus(unsigned int maxcpus) { }
 #endif
 
+#ifdef CONFIG_MANGLE_BOOTARGS
+static void __init mangle_bootargs(char *command_line)
+{
+	char *rootdev;
+	char *rootfs;
+
+	rootdev = strstr(command_line, "root=/dev/mtdblock");
+
+	if (rootdev)
+		strncpy(rootdev, "mangled_rootblock=", 18);
+
+	rootfs = strstr(command_line, "rootfstype");
+
+	if (rootfs)
+		strncpy(rootfs, "mangled_fs", 10);
+
+}
+#else
+static void __init mangle_bootargs(char *command_line)
+{
+}
+#endif
+
 /*
  * We need to store the untouched command line for future reference.
  * We also need to store the touched command line since the parameter
@@ -957,6 +980,7 @@ asmlinkage __visible void __init __no_sanitize_address start_kernel(void)
 	pr_notice("%s", linux_banner);
 	early_security_init();
 	setup_arch(&command_line);
+	mangle_bootargs(command_line);
 	setup_boot_config();
 	setup_command_line(command_line);
 	setup_nr_cpu_ids();
-- 
2.17.1

